File: Exp 1 Sorting Agorithms/counting_sort.cpp
#include <iostream>
#include <vector>
using namespace std;

// Function to insert the array elements
void insert_array_elements(int *arr, int size)
{
    cout << "Enter the elements of the array" << endl;
    for (int i = 0; i < size; i++)
    {
        cout << "- Element " << i + 1 << ": ";
        cin >> arr[i];
    }
}

// Function to display the array
void display_array(int *arr, int size)
{
    for (int i= 0; i < size; i++) cout << arr[i] << " ";
    cout << endl;
}

// Counting Sort function
void counting_sort(int arr[], int size)
{
    int maxval = 0;
    for (int i = 0; i < size; i++) maxval = max(maxval, arr[i]);

    vector<int> count(maxval + 1, 0);
    for (int i = 0; i < size; i++) count[arr[i]]++;
    for (int i = 1; i <= maxval; i++) count[i] += count[i - 1];

    int sorted[size];
    for (int i = size - 1; i >= 0; i--) {
        sorted[count[arr[i]] - 1] = arr[i];
        count[arr[i]]--;
    }

    cout << endl << "Counting Sorted array: ";
    display_array(sorted,size);
}


int main(){
    // Creating the array
    int size;
    cout << "Enter the size of the array: ";
    cin >> size;

    // Insert the elements in the array
    int arr[size];
    insert_array_elements(arr, size);

    // Display the created array
    cout << endl << "Your array: ";
    display_array(arr, size);

    // Counting Sort the array and then display it
    counting_sort(arr, size);

    return 0;
}

File: Exp 1 Sorting Agorithms/counting_sort.py
def insert_array_elements(size):
    """Function to insert the array elements"""
    print("Enter the elements of the array")
    arr = []
    for i in range(size):
        element = int(input(f"- Element {i + 1}: "))
        arr.append(element)
    return arr


def display_array(arr):
    """Function to display the array"""
    print(" ".join(map(str, arr)))


def counting_sort(arr):
    """Counting Sort function"""
    size = len(arr)
    maxval = max(arr)
    
    count = [0] * (maxval + 1)
    for num in arr:
        count[num] += 1
    
    for i in range(1, maxval + 1):
        count[i] += count[i - 1]
    
    sorted_arr = [0] * size
    for i in range(size - 1, -1, -1):
        sorted_arr[count[arr[i]] - 1] = arr[i]
        count[arr[i]] -= 1
    
    print("\nCounting Sorted array: ", end="")
    display_array(sorted_arr)


def main():
    # Creating the array
    size = int(input("Enter the size of the array: "))
    
    # Insert the elements in the array
    arr = insert_array_elements(size)
    
    # Display the created array
    print("\nYour array: ", end="")
    display_array(arr)
    
    # Counting Sort the array and then display it
    counting_sort(arr)


if __name__ == "__main__":
    main()
